### PR TITLE
Fix custom editors in wxGTK wxDataViewCtrl for GTK+2

### PR DESCRIPTION
**Pull request by vadz**
The important commit here is the [second one](https://github.com/wxWidgets/wxWidgets/commit/af8ba52fb0fad2a353b4fb85c43546144bb8002c), fixing [this ticket](http://trac.wxwidgets.org/ticket/17686).

Paul, if you could please check that I am not doing anything particularly stupid here, it would be appreciated. I still have no idea why using `GtkBin` worked with GTK+ 3 but not GTK+ 2, so it's not like I _really_ understand what's going on here... TIA!

### CODE REVIEW

**Review by paulcor — COMMENTED** [review]
GTK+ 3.8 added support for using `GtkBin` like this. `GtkHBox` is deprecated, you could use `GtkBox` instead, although it doesn't work quite right with GTK+ < 2.18 (the edit window appears down at the bottom). Might be acceptable though.

**Comment by vadz** [comment]
Thanks for looking at this!

So can we still continue using `GTKHBox` for now or should I really have 2 different `GtkWxCellEditorBin` implementations, one based on it and another one using `GtkBin`, and create the right one during run-time?

Alternatively, perhaps I could try to make things work with `GtkBin`, if there is no fundamental reason for them not to with GTK+2. I'm a bit dubious about it because I tried doing basically the same thing as it does in 3.8 with overriding `size_allocate` (and `size_request`) vfuncs in `GtkWxCellEditorBin` and forwarding them to the child, but this still didn't work. Maybe I just did something stupid though...

Using `GtkHBox` would still be by far the simplest solution, of course. And if/when it finally disappears in GTK+ N, it would still be simpler to choose between using it and `GtkBin` at compile-time (with checks comparing the major version with `N`) than now at run-time. So maybe we can keep it like this?

> GTK+ 3.8 added support for using `GtkBin` like this

Just to help me to avoid bothering you the next time: how do you know of stuff like this? Are there any secret GTK+ developers channels on the dark web where such things are imparted to the initiated? I could find the relevant change now (I think it's [this commit](https://github.com/GNOME/gtk/commit/3563d11fc6f)) but it would have been very difficult (and obviously too much so for me) to find it without prior knowledge. Is there a useful change log for GTK+ anywhere?

**Comment by paulcor** [comment]
Below are two possible solutions, neither of which is a thing of beauty.

I didn't know about the `GtkBin` changes until you asked me to look at this. I spent a little time digging, compared `gtk/gtkbin.c` from GTK+2 and GTK+3, then did a blame on that file to find out when the changes happened.

BTW, `"editing-canceled"` was added in GTK+ 2.20. Earlier versions spit out a warning:
```
GLib-GObject-WARNING **: g_object_class_override_property: Can't find property to override for 'GtkWxCellEditorBin::editing-canceled'
```

```diff
diff --git a/src/gtk/dataview.cpp b/src/gtk/dataview.cpp
index 6e102ae7c4..fa238f1565 100644
--- a/src/gtk/dataview.cpp
+++ b/src/gtk/dataview.cpp
@@ -1156,7 +1156,13 @@ extern "C" {
 
 struct GtkWxCellEditorBin
 {
+#ifdef __WXGTK4__
     GtkBin parent;
+#else
+    wxGCC_WARNING_SUPPRESS(deprecated-declarations)
+    GtkHBox parent;
+    wxGCC_WARNING_RESTORE()
+#endif
 
     // The actual user-created editor.
     wxWindow* editor;
@@ -1182,9 +1188,16 @@ gtk_wx_cell_editor_bin_get_type()
 
     if ( !cell_editor_bin_type )
     {
+#ifdef __WXGTK4__
+        const guint16 classSize = sizeof(GtkBinClass);
+#else
+        wxGCC_WARNING_SUPPRESS(deprecated-declarations)
+        const guint16 classSize = sizeof(GtkHBoxClass);
+        wxGCC_WARNING_RESTORE()
+#endif
         const GTypeInfo cell_editor_bin_info =
         {
-            sizeof (GtkBinClass),
+            classSize,
             NULL, /* base_init */
             NULL, /* base_finalize */
             gtk_wx_cell_editor_bin_class_init,
@@ -1196,7 +1209,14 @@ gtk_wx_cell_editor_bin_get_type()
             NULL
         };
 
-        cell_editor_bin_type = g_type_register_static( GTK_TYPE_BIN,
+#ifdef __WXGTK4__
+        const GType baseType = GTK_TYPE_BIN;
+#else
+        wxGCC_WARNING_SUPPRESS(deprecated-declarations)
+        const GType baseType = GTK_TYPE_HBOX;
+        wxGCC_WARNING_RESTORE()
+#endif
+        cell_editor_bin_type = g_type_register_static( baseType,
             "GtkWxCellEditorBin", &cell_editor_bin_info, (GTypeFlags)0 );
 
 
```

```diff
diff --git a/src/gtk/dataview.cpp b/src/gtk/dataview.cpp
index 6e102ae7c4..18d073dfa9 100644
--- a/src/gtk/dataview.cpp
+++ b/src/gtk/dataview.cpp
@@ -1222,8 +1222,40 @@ gtk_wx_cell_editor_bin_init(GTypeInstance* instance, void*)
     bin->editor = NULL;
 }
 
+#ifndef __WXGTK4__
+static void cell_editor_size_allocate(GtkWidget* widget, GtkAllocation* a)
+{
+    gtk_widget_set_allocation(widget, a);
+    gtk_widget_size_allocate(gtk_bin_get_child(GTK_BIN(widget)), a);
+}
+
+#ifdef __WXGTK3__
+static void cell_editor_get_preferred_height(GtkWidget* widget, int* min, int* nat)
+{
+    gtk_widget_get_preferred_height(gtk_bin_get_child(GTK_BIN(widget)), min, nat);
+}
+#else
+static void cell_editor_size_request(GtkWidget* widget, GtkRequisition* req)
+{
+    gtk_widget_size_request(gtk_bin_get_child(GTK_BIN(widget)), req);
+}
+#endif
+#endif // !GTK4
+
 static void gtk_wx_cell_editor_bin_class_init(void* klass, void*)
 {
+#ifndef __WXGTK4__
+    if (gtk_check_version(3,8,0))
+    {
+        GtkWidgetClass* widget_class = GTK_WIDGET_CLASS(klass);
+        widget_class->size_allocate = cell_editor_size_allocate;
+#ifdef __WXGTK3__
+        widget_class->get_preferred_height = cell_editor_get_preferred_height;
+#else
+        widget_class->size_request = cell_editor_size_request;
+#endif
+    }
+#endif // !GTK4
     GObjectClass* const oclass = G_OBJECT_CLASS(klass);
 
     oclass->set_property = gtk_wx_cell_editor_bin_set_property;
```

**Comment by vadz** [comment]
OK, thanks again, I think that if we collect all the definitions in one place, the first approach is not too bad -- see the updated commit. And I prefer doing it like this as it will be simpler to just remove the `#else` branch of the code when GTK+4 becomes the only/least supported version.

**Comment by vadz** [comment]
Thanks for the comments, you're totally right about using GTK+ types before the library initialization, I wonder why/how did it work in my testing. Fixed now.

File: src/gtk/dataview.cpp [hunk 1]
```diff
@@ -1152,9 +1152,33 @@ extern "C" {
 #define GTK_TYPE_WX_CELL_EDITOR_BIN               (gtk_wx_cell_editor_bin_get_type ())
 #define GTK_WX_CELL_EDITOR_BIN(obj)               (G_TYPE_CHECK_INSTANCE_CAST ((obj), GTK_TYPE_WX_CELL_EDITOR_BIN, GtkWxCellEditorBin))
 
+// In GTK+ < 3.8 GtkBin can't be used as a widget parent without defining our
```
Comment (paulcor) [inline src/gtk/dataview.cpp]: "widget parent" is a bit confusing here. Maybe something like "base type" or "parent class".

File: src/gtk/dataview.cpp [hunk 2]
```diff
@@ -1152,9 +1152,33 @@ extern "C" {
...
+// own size_allocate and related (either size_request for GTK+ 2 or
+// get_preferred_height for GTK+ 3) vfuncs, so we use GtkHBox instead. But in
+// GTK+ 4, GtkHBox is removed, so we do use GtkBin with it.
+#ifdef __WXGTK4__
+    typedef GtkBin GtkWxCellEditorBinParent;
+    typedef GtkBinClass GtkWxCellEditorBinParentClass;
+
+    // Notice that this can't be just a (const) variable as it GTK+ types are
```
Comment (paulcor) [inline src/gtk/dataview.cpp]: "as it GTK+" typo?

**Review by paulcor — APPROVED** [review]

**Comment by vadz** [comment]
The number of typos I can make in a simple patch is truly amazing, thanks for noticing all (hopefully?) of them -- fixed now and I agree that "Base" is better than "Parent" here.

**Pull request merged**